### ISSUE TITLE
elfeed-tube-add-feeds sometimes returns incomplete results

### ISSUE THREAD

**Issue by Thaodan** [karthink/elfeed-tube#35]
I noticed that elfeed-tube-add-feeds sometimes returns incomplete results.
Meaning that the query contains multiple items such as e.g. keywords but only returns results for a  few of them.
Repeating the query with the remaining missing items finds them.

**Comment by karthink** [comment]
> I noticed that elfeed-tube-add-feeds sometimes returns incomplete results.
> Meaning that the query contains multiple items such as e.g. keywords but only returns results for a  few of them.
> Repeating the query with the remaining missing items finds them.

Right now there are no Invidious servers providing APIs, so elfeed-tube is broken -- it can only retrieve transcripts, because it doesn't use Invidious for that.

I think the Invidious project is more or less dead at this point, so the only recourse is to reimplement the missing features using the user's local yt-dlp application, for which PRs are welcome.

**Comment by Thaodan** [comment]
karthink ***@***.***> writes:

>> I noticed that elfeed-tube-add-feeds sometimes returns incomplete results.
>> Meaning that the query contains multiple items such as e.g. keywords but only returns results for a  few of them.
>> Repeating the query with the remaining missing items finds them.
>
> Right now there are no Invidious servers providing APIs, so
> elfeed-tube is broken -- it can only retrieve transcripts, because it
> doesn't use Invidious for that.

There are some providing API's. I think the issue has been that instead
of rate-limitting they gone ahead to block the API fully.
There are always ways to scrape of pages unstead...

> I think the Invidious project is more or less dead at this point, so the only recourse is to reimplement the missing features using the user's local yt-dlp application, for which PRs are welcome.

I don't know there are some more servers coming but I don't know that
much about the project.

Some kind of python <-> elisp glue code between yt-dlp could help in
this context.

**Comment by karthink** [comment]
>>> I noticed that elfeed-tube-add-feeds sometimes returns incomplete
>>> results.  Meaning that the query contains multiple items such as
>>> e.g. keywords but only returns results for a few of them.  Repeating
>>> the query with the remaining missing items finds them.
>>
>> Right now there are no Invidious servers providing APIs, so
>> elfeed-tube is broken -- it can only retrieve transcripts, because it
>> doesn't use Invidious for that.
>
> There are some providing API's. I think the issue has been that instead
> of rate-limitting they gone ahead to block the API fully.
> There are always ways to scrape of pages unstead...

Scraping youtube pages is unsustainable, I would have to keep updating
elfeed-tube every time youtube changes something, which is often.  In
the years since elfeed-tube was written, I've had to update the fetching
logic only once, and that was because of changes to the Invidious API,
which are rare.  The yt-dlp CLI args are also similarly stable.

So it's better to let yt-dlp or Invidious handle this part.

>> I think the Invidious project is more or less dead at this point, so
>> the only recourse is to reimplement the missing features using the
>> user's local yt-dlp application, for which PRs are welcome.
>
> I don't know there are some more servers coming but I don't know that
> much about the project.

There might be in the future, but it will be good to have a redundant
option like yt-dlp.

> Some kind of python <-> elisp glue code between yt-dlp could help in
> this context.

This is difficult and also unnecessary, we can just call the yt-dlp
process to get what we want.  The main issue is writing the aio-defun
replacements in elfeed-tube.

**Comment by Thaodan** [comment]
karthink ***@***.***> writes:

>> Some kind of python <-> elisp glue code between yt-dlp could help in
>> this context.
>
> This is difficult and also unnecessary, we can just call the yt-dlp
> process to get what we want.  The main issue is writing the aio-defun
> replacements in elfeed-tube.

Parsing shell command is inherently more fragile unless the program is
intentionally supports it.

I wasn't aware that there wasn't a direct bridge for doing such a thing
outside of the Emacs-Python native module.
yt-dlp seems to have a json interface which could be useful to exstract
information.

**Comment by karthink** [comment]
>>> Some kind of python <-> elisp glue code between yt-dlp could help in
>>> this context.
>>
>> This is difficult and also unnecessary, we can just call the yt-dlp
>> process to get what we want.  The main issue is writing the aio-defun
>> replacements in elfeed-tube.
>
> Parsing shell command is inherently more fragile unless the program is
> intentionally supports it.

We won't be parsing shell command output.  yt-dlp usually writes the
metadata to a file, and elfeed-tube will read the file.  I don't know if
the information we need can even be retrieved via stdout.

> I wasn't aware that there wasn't a direct bridge for doing such a thing
> outside of the Emacs-Python native module.
> yt-dlp seems to have a json interface which could be useful to exstract
> information.

Perhaps.  We need the video duration, description and thumbnail.
Transcripts are obtained directly from the Youtube page so we don't need
yt-dlp for that.

**Comment by Thaodan** [comment]
karthink ***@***.***> writes:

>>>> Some kind of python <-> elisp glue code between yt-dlp could help in
>>>> this context.
>>>
>>> This is difficult and also unnecessary, we can just call the yt-dlp
>>> process to get what we want.  The main issue is writing the aio-defun
>>> replacements in elfeed-tube.
>>
>> Parsing shell command is inherently more fragile unless the program is
>> intentionally supports it.
>
> We won't be parsing shell command output.  yt-dlp usually writes the
> metadata to a file, and elfeed-tube will read the file.  I don't know if
> the information we need can even be retrieved via stdout.

The json output can be used similarly how it is done for LSP
integration this should be much faster than writing to disk first.

**Comment by karthink** [comment]
>> We won't be parsing shell command output.  yt-dlp usually writes the
>> metadata to a file, and elfeed-tube will read the file.  I don't know if
>> the information we need can even be retrieved via stdout.
>
> The json output can be used similarly how it is done for LSP
> integration this should be much faster than writing to disk first.

The question is not if it is desirable, but if it is possible.  yt-dlp
can write some things to stdout (as JSON or plain text) and cannot write
other things.  When I last checked, transcripts were one of the things
it could only output to (VTT or equivalent) files.  So the question is
if it can provide the duration, description and thumbnail url as JSON.

**Comment by Thaodan** [comment]
karthink ***@***.***> writes:

>>> We won't be parsing shell command output.  yt-dlp usually writes the
>>> metadata to a file, and elfeed-tube will read the file.  I don't know if
>>> the information we need can even be retrieved via stdout.
>>
>> The json output can be used similarly how it is done for LSP
>> integration this should be much faster than writing to disk first.
>
> The question is not if it is desirable, but if it is possible.  yt-dlp
> can write some things to stdout (as JSON or plain text) and cannot write
> other things.  When I last checked, transcripts were one of the things
> it could only output to (VTT or equivalent) files.  So the question is
> if it can provide the duration, description and thumbnail url as JSON.

- Individual can printed usig the keyword and print e.g. for description:
  - `yt-dlp --print description`
  - These can be taken as is.
- Print json works like this:
  `yt-dlp --dump-json`
 - This can be processed much easier but fetches all metadata.
 - The yt-dlp manual says json metadata can also individually retrieved
   using output templates but I could not get those working.
 - The json output also contains the available captions. These can than
   be retrieved using url-retrieve or alike. From there we can retrieve
   the captions in any format youtube supports.
  
Since the cost of the invocation of yt-dlp is the one which slows down
processing speed using the json metadata should be faster. Batch
processing multiple requests for metadata could also improve speed.


On a side note: It is possible to embed video chapters into the video
output mpv receives using --embed-metadata. I think that's a very useful
in the context of caption since then it's easier to see visually where
you are in the video.

**Comment by karthink** [comment]
>> The question is not if it is desirable, but if it is possible.  yt-dlp
>> can write some things to stdout (as JSON or plain text) and cannot write
>> other things.  When I last checked, transcripts were one of the things
>> it could only output to (VTT or equivalent) files.  So the question is
>> if it can provide the duration, description and thumbnail url as JSON.
>
> - Individual can printed usig the keyword and print e.g. for description:
>   - `yt-dlp --print description`
>   - These can be taken as is.
> - Print json works like this:
>   `yt-dlp --dump-json`
>  - This can be processed much easier but fetches all metadata.
>  - The yt-dlp manual says json metadata can also individually retrieved
>    using output templates but I could not get those working.
>  - The json output also contains the available captions. These can than
>    be retrieved using url-retrieve or alike. From there we can retrieve
>    the captions in any format youtube supports.

Thanks for the research!  It looks like the easy option is a single call
to `yt-dlp --dump-json`.  We don't need yt-dlp to retrieve the
transcript right now, elfeed-tube handles that without Invidious.

> Since the cost of the invocation of yt-dlp is the one which slows down
> processing speed using the json metadata should be faster. Batch
> processing multiple requests for metadata could also improve speed.

Yeah, we'll have to set up a process queue with a maximum number of
simultaneous processes.

> On a side note: It is possible to embed video chapters into the video
> output mpv receives using --embed-metadata. I think that's a very useful
> in the context of caption since then it's easier to see visually where
> you are in the video.

- mpv has done this automatically for me since 2018.  Running
  `mpv https://youtube.com/...` shows the chapters with titles in the
  video timeline.  I think mpv calls yt-dlp with `--embed-metadata`
  internally.
  
- elfeed-tube includes the chapter markers in the transcripts (and
  integrates the chapters with imenu) if everything is working fine.  It
  gets these timestamps and chapter names from the description, so if
  the description is not available there will be no chapters in the
  transcript.  This is the case right now as the Invidious API is not
  available.

**Comment by Thaodan** [comment]
karthink ***@***.***> writes:

>>> The question is not if it is desirable, but if it is possible.  yt-dlp
>>> can write some things to stdout (as JSON or plain text) and cannot write
>>> other things.  When I last checked, transcripts were one of the things
>>> it could only output to (VTT or equivalent) files.  So the question is
>>> if it can provide the duration, description and thumbnail url as JSON.
>>
>> - Individual can printed usig the keyword and print e.g. for description:
>>   - `yt-dlp --print description`
>>   - These can be taken as is.
>> - Print json works like this:
>>   `yt-dlp --dump-json`
>>  - This can be processed much easier but fetches all metadata.
>>  - The yt-dlp manual says json metadata can also individually retrieved
>>    using output templates but I could not get those working.
>>  - The json output also contains the available captions. These can than
>>    be retrieved using url-retrieve or alike. From there we can retrieve
>>    the captions in any format youtube supports.
>
> Thanks for the research!  It looks like the easy option is a single call
> to `yt-dlp --dump-json`.  We don't need yt-dlp to retrieve the
> transcript right now, elfeed-tube handles that without Invidious.

With yt-dlp we would know the possible transcripts such as different
formats and languages at least.

>> Since the cost of the invocation of yt-dlp is the one which slows down
>> processing speed using the json metadata should be faster. Batch
>> processing multiple requests for metadata could also improve speed.
>
> Yeah, we'll have to set up a process queue with a maximum number of
> simultaneous processes.
>
>> On a side note: It is possible to embed video chapters into the video
>> output mpv receives using --embed-metadata. I think that's a very useful
>> in the context of caption since then it's easier to see visually where
>> you are in the video.
>
> - mpv has done this automatically for me since 2018.  Running
>   `mpv https://youtube.com/...` shows the chapters with titles in the
>   video timeline.  I think mpv calls yt-dlp with `--embed-metadata`
>   internally.

Oh yeah I didn't notice that. I first thought some videos not having
chapers was the because the option wasn't passed.

Is it possible to save the point in time you were when dealing with
these streamed videos in mpv?

**Comment by karthink** [comment]
>> Thanks for the research!  It looks like the easy option is a single call
>> to `yt-dlp --dump-json`.  We don't need yt-dlp to retrieve the
>> transcript right now, elfeed-tube handles that without Invidious.
>
> With yt-dlp we would know the possible transcripts such as different
> formats and languages at least.

As I remember I think elfeed-tube already finds the available languages
and fetches the transcript in your specified language of choice.  You
can provide a priority order for the languages it should try.  I'll have
to look at the code to be sure.  The user option is
elfeed-tube-captions-languages.

>>> On a side note: It is possible to embed video chapters into the video
>>> output mpv receives using --embed-metadata. I think that's a very useful
>>> in the context of caption since then it's easier to see visually where
>>> you are in the video.
>>
>> - mpv has done this automatically for me since 2018.  Running
>>   `mpv https://youtube.com/...` shows the chapters with titles in the
>>   video timeline.  I think mpv calls yt-dlp with `--embed-metadata`
>>   internally.
>
> Oh yeah I didn't notice that. I first thought some videos not having
> chapers was the because the option wasn't passed.

There is almost no data exchanged between elfeed-tube and mpv -- it's
only the caption timestamps (for seeking).  Any video metadata mpv uses
is all fetched internally, via mpv's own yt-dlp integration.

> Is it possible to save the point in time you were when dealing with
> these streamed videos in mpv?

Yes, mpv can save your place in videos, see the quit-watch-later mpv
command (mpv's man page explains).

Using mpv's lua scripts, you can do a lot more.  For example, you can
maintain an undo-list across all the seek actions you performed in the
video, and get infinite undo/redo.  Other examples include automatically
skipping sponsored content, clipping videos, downloading them and more.
To be clear -- these are all mpv features and plugins, there's no
connection to elfeed-tube or elfeed.

**Comment by Thaodan** [comment]
karthink ***@***.***> writes:

>>> Thanks for the research!  It looks like the easy option is a single call
>>> to `yt-dlp --dump-json`.  We don't need yt-dlp to retrieve the
>>> transcript right now, elfeed-tube handles that without Invidious.
>>
>> With yt-dlp we would know the possible transcripts such as different
>> formats and languages at least.
>
> As I remember I think elfeed-tube already finds the available languages
> and fetches the transcript in your specified language of choice.  You
> can provide a priority order for the languages it should try.  I'll have
> to look at the code to be sure.  The user option is
> elfeed-tube-captions-languages.
>

OK I guess then it depends on what is easier maintaining the feature or
using ytdl to retrieve possible captions.

>> Is it possible to save the point in time you were when dealing with
>> these streamed videos in mpv?
>
> Yes, mpv can save your place in videos, see the quit-watch-later mpv
> command (mpv's man page explains).

I have save-position-on-quit=yes enabled I assume it just doesn't work
for yt-dlp.

> Using mpv's lua scripts, you can do a lot more.  For example, you can
> maintain an undo-list across all the seek actions you performed in the
> video, and get infinite undo/redo.  Other examples include automatically
> skipping sponsored content, clipping videos, downloading them and more.
> To be clear -- these are all mpv features and plugins, there's no
> connection to elfeed-tube or elfeed.

I know that part I just wasn't sure if mpv is somehow invoked with save
position or a like disabled.

**Comment by karthink** [comment]
>> As I remember I think elfeed-tube already finds the available languages
>> and fetches the transcript in your specified language of choice.  You
>> can provide a priority order for the languages it should try.  I'll have
>> to look at the code to be sure.  The user option is
>> elfeed-tube-captions-languages.
>>
>
> OK I guess then it depends on what is easier maintaining the feature or
> using ytdl to retrieve possible captions.

This feature has no dependencies so it is better to keep it.  Invidious
API access is dead and the user may not have yt-dlp installed --
transcripts in elfeed-tube will still work.

>>> Is it possible to save the point in time you were when dealing with
>>> these streamed videos in mpv?
>>
>> Yes, mpv can save your place in videos, see the quit-watch-later mpv
>> command (mpv's man page explains).
>
> I have save-position-on-quit=yes enabled I assume it just doesn't work
> for yt-dlp.

I don't know the reason then, perhaps some other setting conflicts with
it?  It works for all video sources for me.

>> Using mpv's lua scripts, you can do a lot more.  For example, you can
>> maintain an undo-list across all the seek actions you performed in the
>> video, and get infinite undo/redo.  Other examples include automatically
>> skipping sponsored content, clipping videos, downloading them and more.
>> To be clear -- these are all mpv features and plugins, there's no
>> connection to elfeed-tube or elfeed.
>
> I know that part I just wasn't sure if mpv is somehow invoked with save
> position or a like disabled.

mpv is invoked with playback set to the transcript position where you
click in the elfeed-tube buffer.  If you run elfeed-tube-mpv as a
command and not by clicking on the transcript, it launches with mpv's
default settings, which might involve resuming playback from a saved
position via the quit-watch-later setting.

**Comment by shimeike** [comment]
> Right now there are no Invidious servers providing APIs, so elfeed-tube is broken 

Would it make sense to add a note to this effect at the top of the project Readme? Might save others the time I just spent debugging my configuration for feeds no longer updating and coming to this conclusion ... Unless I missed something, this was the only place I found anything indicating elfeed-tube being broken at present. 

I've hacked together some rough-as-guts `yt-dlp` based stuff to (synchronously) get "description", "duration", and "thumbnail url" strings. Any hints on how I could bypass the existing invidious-based stuff and inject these data into the existing entry modification infrastructure to allow on-demand, synchronous elfeed-entry updates with this data? (The `aio` stuff is a bit beyond me at present.)

Thanks for this package, though! Really enjoyed using it whilst invidious was working.

**Comment by karthink** [comment]
> Would it make sense to add a note to this effect at the top of the project Readme? Might save others the time I just spent debugging my configuration for feeds no longer updating and coming to this conclusion

Yeah.  Would you like to add a note?  It doesn't have to be fancy, just a line at the top of the README stating that thumbnails and descriptions are currently not available because of API changes.

> Any hints on how I could bypass the existing invidious-based stuff and inject these data into the existing entry modification infrastructure to allow on-demand, synchronous elfeed-entry updates with this data?

Sorry, I currently don't have the time to guide someone through both aio and elfeed-tube.  A few notes

**Comment by shimeike** [comment]
> Yeah. Would you like to add a note? It doesn't have to be fancy, just a line at the top of the README stating that thumbnails and descriptions are currently not available because of API changes.

Ok, https://github.com/karthink/elfeed-tube/pull/38

> Sorry, I currently don't have the time to guide someone through both aio and elfeed-tube. I will add a couple of high-level notes later today, but explaining in detail will have to wait.

Oh, sorry - I was definitely not requesting full explanations of either `aio` or `elfeed-tube`! I'm happy to keep plugging away as it's a good learning exercise.

**Comment by karthink** [comment]
@shimeike I looked through the code and there's basically a single function that needs to be rewritten (for basic usage): `elfeed-tube--fetch-desc`.  This function should accept a cl-struct of type `elfeed-entry` corresponding to a youtube feed entry, and return a plist of the following form:

``` emacs-lisp
`(:length ,length-seconds :thumb ,thumb :desc ,desc-html :chaps ,chapters)
```

`thumb` is a url to the thumbnail image, `desc-html` is the video description as html, and `chapters` is an alist of timestamps (I think)  extracted from the description using `elfeed-tube--get-chapters`.  See `elfeed-tube--parse-desc`, it does most of this work for you.

--------

`elfeed-tube--fetch-desc` is an `aio-defun`, so it will basically have to wrap a call to yt-dlp in an `aio` promise.  The call to yt-dlp itself is simple, except in that it requires creating and deleting tmpfiles because yt-dlp does not output the data we need to standard output.  I've written a version of this in a different context before:

``` emacs-lisp
(defun my/gptel-youtube-metadata (callback url)
  (let* ((video-id
          (and (string-match
                (concat
                 "^\\(?:http\\(?:s?://\\)\\)?\\(?:www\\.\\)?\\(?:youtu\\(?:\\(?:\\.be\\|be\\.com\\)/\\)\\)"
                 "\\(?:watch\\?v=\\)?" "\\([^?&]+\\)")
                url)
               (match-string 1 url)))
         (dir (file-name-concat temporary-file-directory "yt-dlp" video-id)))
    (if (file-directory-p dir) (delete-directory dir t))
    (make-directory dir t)
    (let ((default-directory dir) (idx 0)
          (data (list :description nil :transcript nil)))
      (make-process :name "yt-dlp"
                    :command `("yt-dlp" "--write-description" "--skip-download" "--output" "video" ,url)
                    :sentinel (lambda (proc status)
                                (cl-incf idx)
                                (let ((default-directory dir))
                                  (when (file-readable-p "video.description")
                                    (plist-put data :description
                                               (with-temp-buffer
                                                 (insert-file-contents "video.description")
                                                 (buffer-string)))))
                                (when (= idx 2)
                                    (funcall callback (gptel--json-encode data))
                                    (delete-directory dir t))))
      (make-process :name "yt-dlp"
                    :command `("yt-dlp" "--skip-download" "--write-auto-subs" "--sub-langs"
                               "en,-live_chat" "--convert-subs" "srt" "--output" "video" ,url)
                    :sentinel (lambda (proc status)
                                (cl-incf idx)
                                (let ((default-directory dir))
                                  (when (file-readable-p "video.en.srt")
                                    (plist-put data :transcript
                                               (with-temp-buffer
                                                 (insert-file-contents "video.en.srt")
                                                 (buffer-string)))))
                                (when (= idx 2)
                                    (funcall callback (gptel--json-encode data))
                                    (delete-directory dir t)))))))
```

This version fetches the transcripts too, which we don't want since transcript fetching is still working well in elfeed-tube.

------

Two additional points:

1. Instead of replacing `elfeed-tube--fetch-desc` entirely, I would create a top level user option to control if elfeed-tube tries to use yt-dlp or invidious, defaulting to yt-dlp if it's available on the system.  The value of this variable will determine which of the two implementations of `elfeed-tube--fetch-desc` is called.

2. Many other features of elfeed-tube continue to be non-functional: searching for youtube feeds, backfilling historical videos from before the rolling feed start date, etc.  Perhaps we can do some of this using yt-dlp too in the future, I haven't investigated.

-------

Hope this gives you or anyone else interested in fixing elfeed-tube a feasible starting point without having to read the source for the whole package.

**Comment by shimeike** [comment]
> This function should accept a cl-struct of type `elfeed-entry` corresponding to a youtube feed entry, and return a plist of the following form:
> 
> `(:length ,length-seconds :thumb ,thumb :desc ,desc-html :chaps ,chapters)
> 
> `thumb` is a url to the thumbnail image, `desc-html` is the video description as html, and `chapters` is an alist of timestamps (I think) extracted from the description using `elfeed-tube--get-chapters`. See `elfeed-tube--parse-desc`, it does most of this work for you.

@karthink thanks! This is "more than exactly" what I was looking for!

Until now I have been playing around in an entirely separate namespace (systematically backwards from your main entry points) as I had zero confidence of doing anything that would be useful beyond my specific use-case. But maybe with these hints I can try to do a bit better.

The following is what I had (obviously wrong data structures based on your feedback) for the `yt-dlp` part. I was only hoping to replicate `C-u F` behaviour from a single `elfeed-entry` buffer.

```emacs-lisp
(defun my/elfeed-tube-getdata (URL)
  "Returns a list containing description, duration, and thumbnail url for video at URL"
  (let* ((json-file (make-temp-file "elfeed-tube-")))
    (with-temp-file json-file
      (insert (shell-command-to-string
               (concat
                "yt-dlp --skip-download --dump-json " URL))))
    (let* ((json-object-type 'hash-table)
           (json-array-type 'list)
           (json-key-type 'string)
           (videodata (json-read-file json-file)))
      (delete-file json-file)
      (list
       (gethash "description" videodata) ; description
       (format-seconds "%.2h:%.2m:%.2s" (gethash "duration" videodata)) ; duration
       (gethash "thumbnail" videodata) ; thumbnail url
       ))))

**Comment by LionyxML** [comment]
Hello there, just blindly shooting in the dark (since I haven't yet dug into the source-code), but do you guys see the yt-dlp command: `yt-dlp --write-thumbnail --skip-download URL` as an option?

This actually downloads the thumbnail in the `.webp` file, while `yt-dlp --write-thumbnail --convert-thumbnails jpg --skip-download URL` also converts it to a `.jpg`.

The downside is that it would have to write the file in some `.cache` or create a temp file to be consumed and deleted when presenting the feed.

**Comment by shimeike** [comment]
So, the below (which can be found [here](https://github.com/shimeike/elfeed-tube/tree/feature/yt-dlp)) seems to fill an entry with thumbnail, duration, and description data (via default `C-u F` keybinding). @karthink's hints were sufficient to just replicate the existing framework with zero `aio` understanding.

Following are my comments/observations:

- As mentioned, I have no idea about `aio` implications of any of the below. (Maybe it's not even utilised with the method I called `yt-dlp`.)
- Not well tested (not even sure how to test - I just went back through a few recent elfeed entries and manually filled them in)
- I have not seen what happens when updating elfeed pulls in 1 or more youtube entries (again, not sure how to test)
- Description is plaintext (this seems to be all that's present in the returned json)
- Haven't looked into chapters, though they didn't seem to be working. Maybe the existing parsing relies on html?
- Image at the url returned by the easy-to-access `thumbnail` key is too large. There is also a `thumbnails` json key that contains smaller images, but I haven't looked into whether they're repeatable across different videos.
- No logging or error handling
- There may be better elisp paradigms for some/all of this

```diff
modified   elfeed-tube.el
@@ -196,6 +196,13 @@ entries that don't have metadata."
   :group 'elfeed-tube
   :type 'boolean)
 
+(defcustom elfeed-tube-use-ytdlp-p t
+  "Whether to use yt-dlp to obtain description, thumbnail, duration.
+
+This is a boolean. Use yt-dlp when t. Otherwise use invidious."
+  :group 'elfeed-tube
+  :type 'boolean)
+
 (defcustom elfeed-tube-captions-sblock-p t
   "Whether sponsored segments should be de-emphasized in transcripts."
   :group 'elfeed-tube
@@ -965,7 +972,37 @@ The result is a plist with the following keys:
                   (setf (car telm) cat))
              finally return captions)))
 
-(aio-defun elfeed-tube--fetch-desc (entry &optional attempts)
+(defun elfeed-tube--fetch-desc (entry &optional attempts)
+  (if elfeed-tube-use-ytdlp-p
+      (elfeed-tube--fetch-desc-ytdlp entry attempts)
+    (elfeed-tube--fetch-desc-invid entry attempts)))
+
+(aio-defun elfeed-tube--fetch-desc-ytdlp (entry &optional attempts)
+  "Returns a list containing description, duration, and thumbnail url for video at URL"
+  (let* ((attempts (or attempts (1+ elfeed-tube--max-retries)))
+         (video-id (elfeed-tube--entry-video-id entry))
+         (json-file (make-temp-file "elfeed-tube-")))
+    (with-temp-file json-file
+      (insert (shell-command-to-string
+               (concat
+                "yt-dlp --skip-download --dump-json " video-id))))
+    (let* ((json-object-type 'hash-table)
+           (json-array-type 'list)
+           (json-key-type 'string)
+           (videodata (json-read-file json-file)))
+      (delete-file json-file)
+      (list
+       :length
+       (gethash "duration" videodata) ; duration
+       :thumb
+       (gethash "thumbnail" videodata) ; thumbnail url possibly too large. Need to get appropriate size.
+       :desc
+       (gethash "description" videodata) ; description - plain test. Need to get html.
+       :chaps
+       (gethash "chapters" videodata)
+       ))))
+
+(aio-defun elfeed-tube--fetch-desc-invid (entry &optional attempts)
   (let* ((attempts (or attempts (1+ elfeed-tube--max-retries)))
          (video-id (elfeed-tube--entry-video-id entry)))
     (when (> attempts 0)
```

**Comment by shimeike** [comment]
> Regarding the `:thumb` part. Do you guys see the yt-dlp command: `yt-dlp --write-thumbnail --skip-download URL` as an option?

Hi @LionyxML, it could be. But, as you point out, it would be a different architecture than the already existing Invidious framework. 'yt-dlp' can also directly print strings for some fields as others have discussed above. I don't know if this, combined with your suggestion, would be more efficient than grabbing the entire json, or if `yt-dlp` just grabs the entire json regardless and then processes it internally.

**Comment by Thaodan** [comment]
> The downside is that it would have to write the file in some `.cache` or create a temp file to be consumed and deleted when presenting the feed.

There's url-cache-directory for that kind of thing for example.

**Comment by shimeike** [comment]
After a bit (actually maybe a lot ...?!) of going around in emacs process-related learning circles, I appear to have something maybe working with `aio`. Again, over at this  ([feature/yt-dlp](https://github.com/shimeike/elfeed-tube/tree/feature/yt-dlp)) branch. The big clue came from [here](https://github.com/skeeto/emacs-aio/issues/19#issuecomment-729660484) and it's astonishing how simple things seem to end up.

Feedback welcome!

Comments from my side:
- Still no html in the description. In fact, at the moment, even blank lines get discarded at some point in the entry-writing process, so the description does not look as nice as it should. I don't think html is possible, but it would be great to retain the plain text formatting.
- I've mapped the existing `thumbnail-size` setting onto arbitrary, but seemingly consistent, thumbnail sizes given in the json. The resulting images are, at the moment, larger than what resulted from the corresponding settings before. There are probably several possibilities to improve this.
- I have only ever used `elfeed-tube` for duration, description, and thumbnail. These are all done via `yt-dlp` and appear to be working. I don't know if other functionality is impacted/broken as I don't have experience with how it worked before.
- `chapters` are currently extracted from the json and returned in the plist. However, I don't know the required data format as it's based on parsing html that I cannot get.
- No `yt-dlp`-specific logging / error handling / retries yet

**Comment by shimeike** [comment]
Pull request #39 created. @karthink let me know if you need this squashed and I'll figure it out.

**Comment by karthink** [comment]
Closing this issue since yt-dlp support has been implemented by @shimeike.

Advanced features like back-filling feeds or searching for Youtube RSS feeds is still broken because Invidious is gone, but that's a separate issue.
